### PR TITLE
feat: huffman

### PR DESCRIPTION
**Pull request by ChillerDragon**

### CODE REVIEW

**Comment by ChillerDragon** [comment]
Resolved conflicts.

**Comment by MilkeeyCat** [comment]
There're a lot of `int`s and `unsigned`s etc.

**Comment by ChillerDragon** [comment]
Oops fixed. That should probably be added to the CI.

File: src/huffman.c [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+	167, 94, 128, 122, 223, 153, 218, 77, 200, 110, 190, 73, 174, 69, 145, 66, 277, 143, 141, 60,
+	136, 53, 180, 57, 142, 57, 158, 61, 166, 112, 152, 92, 26, 22, 21, 28, 20, 26, 30, 21,
+	32, 27, 20, 17, 23, 21, 30, 22, 22, 21, 27, 25, 17, 27, 23, 18, 39, 26, 15, 21,
+	12, 18, 18, 27, 20, 18, 15, 19, 11, 17, 33, 12, 18, 15, 19, 18, 16, 26, 17, 18,
+	9, 10, 25, 22, 22, 17, 20, 16, 6, 16, 15, 20, 14, 18, 24, 335, 1517};
+
+struct HuffmanConstructNode {
+	uint16_t node_id;
+	uint32_t frequency;
+} typedef HuffmanConstructNode;
```
Comment (MilkeeyCat) [inline src/huffman.c:21]: Is it not possible to use just `typedef struct {} Name;`? Also move them in header file.

File: src/huffman.c [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+	// symbol
+	uint32_t bits;
+	uint32_t num_bits;
+
+	// don't use pointers for this. shorts are smaller so we can fit more data into the cache
+	uint16_t leafs[2];
+
+	// what the symbol represents
+	uint8_t symbol;
+} typedef Node;
```
Comment (MilkeeyCat) [inline src/huffman.c:33]: Same here

**Review by MilkeeyCat — CHANGES_REQUESTED** [review]
Also, you have to add a new typedef in `common.h` for `int32_t` and `memset`

File: src/huffman.c (hunk 1 above)
Comment (ChillerDragon) [reply src/huffman.c:21]: I wanted to keep the header file clean and only expose the user relevant api.

Comment (MilkeeyCat) [reply src/huffman.c:21]: Ah, ye, it's not needed, my bad :\

**Comment by ChillerDragon** [comment]
> Also, you have to add a new typedef in `common.h` for `int32_t` and `memset`

Yea I thought about the memset but then I did not really understand that black magic and forgot about it xd.
If it it is not including or linking libc doesn't that mean we need a full implementation of memset not just an empty signature?

**Comment by MilkeeyCat** [comment]
Ye, the user of the library will implement those functions, I started this repo because I needed a lib to parse packets in kernel module. And it's not possible to link glibc in kernel module xd

**Comment by MilkeeyCat** [comment]
Thanks

**Pull request merged**